sklearn/cluster/k_means_.py

```diff
@@ -414,14 +436,20 @@ def _kmeans_single_elkan(X, n_clusters, max_iter=300, init='k-means++',
     centers = np.ascontiguousarray(centers)
     if verbose:
         print('Initialization complete')
-    centers, labels, n_iter = k_means_elkan(X, n_clusters, centers, tol=tol,
+
+    checked_sample_weights = _check_sample_weights(X, sample_weights)
+    centers, labels, n_iter = k_means_elkan(X, checked_sample_weights,
+                                            n_clusters, centers, tol=tol,
                                             max_iter=max_iter, verbose=verbose)
-    inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)
+    sq_distances = (X - centers[labels]) ** 2
+    if sample_weights is not None:
+        sq_distances *= checked_sample_weights[:, np.newaxis]
```

Comment:
Actually, on second thought, the operation is expensive because of a useless broadcasting.
Instead of `n_samples * n_features` multiplications, we can do only `n_samples` multiplications with:
```py
if sample_weights is not None:
    sq_distances = np.sum(sq_distances, axis=1, dtype=np.float64)
    sq_distances *= checked_sample_weights
```

Reply:
You are right that this makes the computation faster if `sample_weights` is _not_ `None`. It obviously doesn't make a difference when no sample weights are passed.

I should clarify that when I said above that _"I cannot make out any statistically significant difference"_ I was referring to the difference between the `weighted_k_means` and `master` branches, not before and after the change in `_kmeans_single_elkan`.

Reply:
Sure, but removing the unnecessary broadcasting should speed up the `sample_weights is not None` case.

Reply:
Definitely, I'm testing it right now

Reply:
Okay, new benchmarks, adding:
```python
w = np.ones(X.shape[0])
```

On branch `weighted_k_means`:
```console
$ python -m timeit -n 100 -s 'from benchmark_k_means import X,km,w' 'km.fit(X,sample_weights=w)'
100 loops, best of 3: 40.1 msec per loop
$ python -m timeit -n 100 -s 'from benchmark_k_means import X,km' 'km.fit(X)'
100 loops, best of 3: 39.4 msec per loop
```

On branch `master`:
```console
$ python -m timeit -n 100 -s 'from benchmark_k_means import X,km' 'km.fit(X)'
100 loops, best of 3: 38 msec per loop
```

All subject to ~ 1 msec variation.